### PR TITLE
Accept array of secrets, not just one

### LINKED ISSUE #770: [FEAT]: Facilitate webhook secret rotation
### Describe the need

The current `verifyAndReceive` function assumes that there is exactly one secret in use at any time:
https://github.com/octokit/webhooks.js/blob/d6286fa324b65716c12ce03efe338faf0e9d2338/src/verify-and-receive.ts#L10-L11
which frustrates secret rotation, which can be a requirement in organizational settings.  In particular, the current design essentially forces either programmer suffering (maintaining multiple webhook objects) or a window in which the sender and receiver will disagree about the (singular) secret value.

It would be ideal if, instead, `verifyAndReceive` accepted a list of secrets and attempted verification against each one.  Then key rotation is straightforward:
1. Generate a new secret.
2. Update the receiver to accept messages MAC'd by this secret.  Wait for things to settle.
3. Update GitHub to send messages with this secret.  Wait for things to settle.
4. Remove the old secret from the receiver.

### SDK Version

_No response_

### API Version

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by nwf-msr**
Resolves #770.

----

### Before the change?

* Only one secret could be specified for a WebHook object.

### After the change?

* A bare secret or an array of secrets are each accepted.  In the latter case, any match will be accepted.

----

### Pull request checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Added the appropriate label for the given change

### Does this introduce a breaking change?

- [x] No

### Pull request type

I do not seem to be able to add labels, but please add `Type: Feature` for me?

----

### CODE REVIEW

**Review by gr2m — CHANGES_REQUESTED** [review]
PR looks great so far! Can you please also update the `### Constructor` options table?

Also can you run `npm run lint:fix` to make the prettier linter happy?

**Comment by gr2m** [comment]
can you look into the failing tests?

Please don't force-push changes, we don't mind a messy commit history in pull requests, we squash & merge them in the end anyway

**Comment by nwf-msr** [comment]
What'd I do to make CodeQL unhappy? :(

**Comment by wolfy1339** [comment]
It's not you, it's CodeQL being uncooperative.

Just ignore it

File: README.md [hunk 1]
```diff
@@ -56,6 +56,18 @@ require("http").createServer(createNodeMiddleware(webhooks)).listen(3000);
 // can now receive webhook events at /api/github/webhooks
 ```
 
+To ease key rotation, one can also feed the `Webhook` constructor an array of
+secrets. Generate a new secret and update your Webhook receiver to accept both
+this new one and the current secret, then update GitHub to send the new secret,
+then update your Webhook receiver to accept only the new secret. Signatures will
+always be made with the first secret.
```
Comment (gr2m) [inline README.md:63]:
```suggestion
then update your Webhook receiver to accept only the new secret. New signatures will
be calculated using the first secret.
```

File: README.md [hunk 2]
```diff
@@ -111,11 +123,14 @@ new Webhooks({ secret /*, transform */ });
         <code>
           secret
         </code>
-        <em>(String)</em>
+        <em>(String or String Array)</em>
```
Comment (gr2m) [inline README.md:126]:
```suggestion
        <em>(String or Array of Strings)</em>
```

File: README.md [hunk 3]
```diff
@@ -111,11 +123,14 @@ new Webhooks({ secret /*, transform */ });
...
       </td>
       <td>
         <strong>Required.</strong>
-        Secret as configured in GitHub Settings.
+        Secret as configured in GitHub Settings.  If an array of strings, an
+        incoming webhook event will be accepted if its signature verifies
+        with any of the secrets given and signatures will always be generated
+        with the first secret.
```
Comment (gr2m) [inline README.md:133]:
```suggestion
        Secret as configured in GitHub Settings.  If set to an array of strings, a
        webhook request will be accepted if its signature verifies
        with any of the given secrets. New signatures will always be
        calculated using the first secret.
```

**Review by gr2m — COMMENTED** [review]
I think the pull request looks good as is.

Good point about the `.sign()` method. Given that the first secret in a given array of secrets has special meaning, I wonder if it would make more sense to keep the secret as is, but add an additional option such as `alternativeSecrets` or maybe some other name that better conveys that these secrets are used for incoming request verification only? 

My understanding is that in your case, there will always only be one other secret. Do you see a use case where more than two secrets would be needed to be provided?

**Comment by nwf-msr** [comment]
> I think the pull request looks good as is.
> 
> Good point about the `.sign()` method. Given that the first secret in a given array of secrets has special meaning, I wonder if it would make more sense to keep the secret as is, but add an additional option such as `alternativeSecrets` or maybe some other name that better conveys that these secrets are used for incoming request verification only?

That makes sense.  I'll take a stab at that approach.

> My understanding is that in your case, there will always only be one other secret. Do you see a use case where more than two secrets would be needed to be provided?

That's true and no, I don't see reason to need more than one active and one retiring secret, at least, at the moment.  I'm tempted to leave `alternativeSecrets` as an array of strings, tho', just in case.  It's not like it's hard or on the hot path most of the time.

**Comment by gr2m** [comment]
maybe call them `alternativeSecretsForVerification` to make clear they won't be used for the `.sign()` method?

**Comment by nwf-msr** [comment]
Belatedly, I've raised https://github.com/octokit/webhooks.js/pull/811 .

**Comment by wolfy1339** [comment]
Instead of opening a separate PR, you could have simply reused this one.

Closing in favour of #811

**Comment by nwf-msr** [comment]
I didn't reuse this one because I was advised not to force-push changes and enough time had elapsed that doing the work on a new checkout seemed wise.

**Comment by wolfy1339** [comment]
When you need to update using a rebase, a force-push can be warranted.
When you simply always squash your changes, it can become harder to review changes since the last time it was reviewed, and we don't mind having many commits.

**Pull request closed without merging**